File: Compiler/src/compiler.ts
import { Parser } from './parser';
import { Tokenizer } from './tokenizer';
import { FunctionToken } from './types/parsedToken';
import { Verifier } from './verifier';

type ConfigType = {
	filename: string;
	content: string;
};

export class Compiler {
	constructor() {}

	run(config: ConfigType) {
		const { filename, content } = config;

		const tokenizer = new Tokenizer(filename, content);
		const parser = new Parser(tokenizer, content, filename);

		const { tokens } = parser.parse();
		const functions = tokens.filter(c => c.type === 'FunctionDeclaration') as FunctionToken[];

		const verifier = new Verifier(filename, functions);
		verifier.verifyBody(tokens);
	}
}

File: Compiler/src/compiler.rs
use std::fs;
use std::path::PathBuf;

use regex::Regex;

use crate::parsers::Parser;
use crate::tokenizer::Tokenizer;
use crate::utils::{ImportedModule, ParsedToken};

pub struct Compiler {
    config: CompilerConfig,
}

#[derive(Clone, Debug)]
pub struct CompilerConfig {
    pub isNode: bool,
    pub es6: bool,
    pub filename: String,
    pub content: String,
}

impl CompilerConfig {
    pub fn new(path: PathBuf, filename: String, isNode: bool, es6: bool) -> Self {
        let content = fs::read_to_string(path.clone())
            .expect(format!("Failed to read file {}", path.display()).as_str());

        Self {
            isNode,
            filename,
            content,
            es6,
        }
    }
}

impl Compiler {
    pub fn new(config: CompilerConfig) -> Self {
        Self { config }
    }

    pub fn run(&mut self) -> (String, Vec<ParsedToken>, Vec<ImportedModule>) {
        let CompilerConfig {
            filename: path,
            content,
            isNode,
            es6,
        } = self.config.clone();

        let regex = Regex::new(r"(.+)/$").unwrap();

        let filename = path.split("/").last().unwrap();
        let result = path.replace(filename, "");
        let folder = regex.replace(result.as_str(), "$1").to_string();

        let tokenizer = Tokenizer::new(path.clone(), content.clone());
        let mut parser = Parser::new(tokenizer, folder, isNode, es6);

        return parser.run();
    }
}
